File: app/apify.py
"""Apify actors chalata hai aur dono sources ka output ek jaise shape me laata hai."""
import asyncio
from typing import Any, Awaitable, Callable

from apify_client import ApifyClientAsync

# bebity/linkedin-jobs-scraper ke liye paid rental chahiye; ye pay-per-event actor free plan par chalta hai.
LINKEDIN_ACTOR = "curious_coder/linkedin-jobs-scraper"
INDEED_ACTOR = "misceres/indeed-scraper"

ProgressFn = Callable[..., Awaitable[None]]


def _normalize_linkedin(job: dict) -> dict:
    return {
        "source": "linkedin",
        "title": job.get("title"),
        "company": job.get("companyName"),
        "location": job.get("location"),
        "postedAt": job.get("postedAt") or job.get("publishedAt"),
        "contractType": job.get("employmentType") or None,
        "experienceLevel": job.get("seniorityLevel") or None,
        "workType": job.get("workplaceType") or None,
        "salary": job.get("salary") or None,
        "url": job.get("link") or job.get("jobUrl"),
        "applyUrl": job.get("applyUrl") or None,
        "applicants": job.get("applicantsCount") or None,
        "description": job.get("descriptionText") or job.get("description"),
    }


def _normalize_indeed(job: dict) -> dict:
    job_type = job.get("jobType")
    return {
        "source": "indeed",
        "title": job.get("positionName"),
        "company": job.get("company"),
        "location": job.get("location"),
        "postedAt": job.get("postedAt") or job.get("postingDateParsed"),
        "contractType": ", ".join(job_type) if isinstance(job_type, list) else job_type,
        "experienceLevel": None,
        "workType": "Remote" if job.get("isRemote") else None,
        "salary": job.get("salary"),
        "url": job.get("url"),
        "applyUrl": job.get("externalApplyLink"),
        "applicants": None,
        "description": job.get("description"),
    }


async def scrape_jobs(
    *,
    title: str,
    location: str,
    country: str,
    limit: int,
    source: str = "both",
    token: str,
    on_progress: ProgressFn | None = None,
) -> list[dict[str, Any]]:
    """Dono job boards ko parallel me scrape karta hai, phir URL par dedupe."""
    client = ApifyClientAsync(token=token)

    async def progress(stage: str, message: str, is_error: bool = False) -> None:
        if on_progress:
            await on_progress(stage, message, is_error)

    async def run_one(actor_id: str, run_input: dict, label: str, normalize) -> list[dict]:
        await progress(label, f"{label} scraper start ho raha hai...")
        try:
            run = await client.actor(actor_id).call(run_input=run_input)
            if run is None or not run.default_dataset_id:
                raise RuntimeError("Apify run ka dataset nahi mila (run fail hua ya timeout)")
            page = await client.dataset(run.default_dataset_id).list_items()
            items = page.items
            await progress(label, f"{label} se {len(items)} jobs mile")
            return [normalize(item) for item in items]
        except Exception as err:  # ek source fail ho to dusri ka data phir bhi dikhao
            await progress(label, f"{label} fail hua: {err}", True)
            return []

    tasks = []

    if source in ("both", "linkedin"):
        tasks.append(
            run_one(
                LINKEDIN_ACTOR,
                {"keywords": title, "location": location, "limitPerSource": limit, "scrapeCompany": False},
                "linkedin",
                _normalize_linkedin,
            )
        )

    if source in ("both", "indeed"):
        tasks.append(
            run_one(
                INDEED_ACTOR,
                {
                    "position": title,
                    "location": location,
                    "country": country,
                    "maxItemsPerSearch": limit,
                    "parseCompanyDetails": False,
                    "saveOnlyUniqueItems": True,
                    "followApplyRedirects": False,
                },
                "indeed",
                _normalize_indeed,
            )
        )

    seen: set[str] = set()
    jobs: list[dict] = []
    for batch in await asyncio.gather(*tasks):
        for job in batch:
            url = job.get("url")
            if url:
                if url in seen:
                    continue
                seen.add(url)
            jobs.append(job)
    return jobs

File: lib/apify.js
import { ApifyClient } from 'apify-client';

// bebity/linkedin-jobs-scraper ke liye paid rental chahiye; ye pay-per-event actor free plan par chalta hai.
const LINKEDIN_ACTOR = 'curious_coder/linkedin-jobs-scraper';
const INDEED_ACTOR = 'misceres/indeed-scraper';

// Har actor ka output shape alag hai, isliye ek common shape me normalize kar rahe hain.
function normalizeLinkedIn(job) {
    return {
        source: 'linkedin',
        title: job.title ?? null,
        company: job.companyName ?? null,
        location: job.location ?? null,
        postedAt: job.postedAt ?? job.publishedAt ?? null,
        contractType: job.employmentType || null,
        experienceLevel: job.seniorityLevel || null,
        workType: job.workplaceType || null,
        salary: job.salary || null,
        url: job.link ?? job.jobUrl ?? null,
        applyUrl: job.applyUrl || null,
        applicants: job.applicantsCount || null,
        description: job.descriptionText ?? job.description ?? null,
    };
}

function normalizeIndeed(job) {
    return {
        source: 'indeed',
        title: job.positionName ?? null,
        company: job.company ?? null,
        location: job.location ?? null,
        postedAt: job.postedAt ?? job.postingDateParsed ?? null,
        contractType: Array.isArray(job.jobType) ? job.jobType.join(', ') : (job.jobType ?? null),
        experienceLevel: null,
        workType: job.isRemote ? 'Remote' : null,
        salary: job.salary ?? null,
        url: job.url ?? null,
        applyUrl: job.externalApplyLink ?? null,
        applicants: null,
        description: job.description ?? null,
    };
}

/**
 * Dono job boards ko parallel me scrape karta hai.
 * onProgress(stage, message) se caller ko live updates milte hain.
 */
export async function scrapeJobs({ title, location, country, limit, source = 'both', token }, onProgress = () => {}) {
    const client = new ApifyClient({ token });

    const runOne = async (actorId, input, label, normalize) => {
        onProgress(label, `Starting the ${label} scraper...`);
        try {
            const run = await client.actor(actorId).call(input);
            const { items } = await client.dataset(run.defaultDatasetId).listItems();
            onProgress(label, `Got ${items.length} jobs from ${label}`);
            return items.map(normalize);
        } catch (err) {
            onProgress(label, `${label} failed: ${err.message}`, true);
            return [];
        }
    };

    const tasks = [];

    if (source === 'both' || source === 'linkedin') {
        tasks.push(runOne(LINKEDIN_ACTOR, {
            keywords: title,
            location,
            limitPerSource: limit,
            scrapeCompany: false,
        }, 'linkedin', normalizeLinkedIn));
    }

    if (source === 'both' || source === 'indeed') {
        tasks.push(runOne(INDEED_ACTOR, {
            position: title,
            location,
            country,
            maxItemsPerSearch: limit,
            parseCompanyDetails: false,
            saveOnlyUniqueItems: true,
            followApplyRedirects: false,
        }, 'indeed', normalizeIndeed));
    }

    const seen = new Set();
    return (await Promise.all(tasks)).flat().filter((j) => {
        if (!j.url) return true;
        if (seen.has(j.url)) return false;
        seen.add(j.url);
        return true;
    });
}
